x/liquidstake/keeper/liquidstake.go

```diff
@@ -719,6 +723,82 @@ func (k Keeper) LiquidUnbond(
 	return completionTime, returnAmount, ubd, nil
 }
 
+// DivideByWeight divide the input amount to delegate between the active validators, specifically checking if they have
+// room to receive more liquid delegations. The leftover that might be left is divided across all validators that can
+// receive it.
+func (k Keeper) DivideByWeight(
+	ctx sdk.Context,
+	avs types.ActiveLiquidValidators,
+	totalInput math.Int,
+	whitelistedValsMap types.WhitelistedValsMap,
+) map[string]math.Int {
+	totalWeight := avs.TotalWeight(whitelistedValsMap)
+	if !totalWeight.IsPositive() {
+		return map[string]math.Int{}
+	}
+
+	totalDelegation := sdk.ZeroInt()
+	delegations := make(map[string]math.Int)
+	inputPerValidator := math.LegacyNewDecFromInt(totalInput).QuoTruncate(math.LegacyNewDecFromInt(totalWeight))
+
+	// loop through all validators and check if the amount of shares they would receive is within the validator specific caps
+	for _, val := range avs {
+		validator, _ := k.stakingKeeper.GetValidator(ctx, val.GetOperator())
+
+		// calculate the shares the input would receive
+		delegation := inputPerValidator.MulInt(val.GetWeight(whitelistedValsMap, true)).TruncateInt()
+		if !delegation.IsPositive() { // continue if the delegation is 0
+			continue
+		}
+		delegationShares := validator.GetDelegatorShares().MulInt(delegation).QuoInt(validator.GetTokens())
+
+		// just delegate if the validator does not exceed any of the validator specific caps
+		if !k.stakingKeeper.CheckExceedsValidatorBondCap(ctx, validator, delegationShares) &&
+			!k.stakingKeeper.CheckExceedsValidatorLiquidStakingCap(ctx, validator, delegationShares, false) {
+			totalDelegation = totalDelegation.Add(delegation)
+			delegations[val.GetOperator().String()] = delegation
+		}
+	}
+
+	if len(delegations) == 0 {
+		return map[string]math.Int{}
+	}
+
+	// the leftover amount that could not be delegated is divided across all validators
+	leftOver := totalInput.Sub(totalDelegation)
+	numValidators := math.NewInt(int64(len(delegations)))
+	for leftOver.IsPositive() {
+		diffPerValidator := leftOver.Quo(numValidators)
+		if leftOver.LT(numValidators) { // if the leftover is less than the number of validators to be delegated, just add it to the first one that can receive it
+			diffPerValidator = leftOver
+		}
+		for strAddr := range delegations {
```

Comment:
this is a ticking bomb for non determinism. do not iterate over maps. both here and https://github.com/persistenceOne/pstake-native/pull/810/files#diff-d71073dfa714aa345a7e072ed4d0626328925c093d70acb85254b12df28eb45dR562